### DISCUSSION TITLE
Compiler Error: Invalid Implicit Conversion from Contract AggregatorV3Interface

### DISCUSSION THREAD

**Discussion by chapman91** [Cyfrin/foundry-full-course-cu#2139] [Q&A] [answered]
I'm encountering a compilation error in my Solidity project when using the Chainlink AggregatorV3Interface. The error message is:

![invalid_implicit_conversion_error](https://github.com/user-attachments/assets/e984afde-8bd7-4470-b380-636d9ef1eac0)

The contract is designed to accept an address of a deployed AggregatorV3Interface during construction and use it for price conversion.

**Code Snippet:**
 
![fundme_cb_ss](https://github.com/user-attachments/assets/de0db4b5-acd9-412a-96c2-481da6a0cd5c)

Steps Taken:

1. Imported the Interface: Imported AggregatorV3Interface from Chainlink's contracts.

2. Constructor: Passed the address of the price feed to the constructor.

3. Used the Interface: Called latestRoundData() from the AggregatorV3Interface.

Questions:

Why am I getting an invalid implicit conversion error when the types seem to match?

How can I ensure the AggregatorV3Interface type is consistent across my codebase?

Is there a specific version of Chainlink's contracts I should be using?

**Comment by EngrPips** [comment]
Hello @chapman91, Please show your `PriceConverter` library.

**Comment by chapman91** [comment]
![PriceConverter_ScreenShot](https://github.com/user-attachments/assets/f9889b75-9b0a-4fa0-a17c-dffd9ca62e49)

**Reply by EngrPips** [reply]
Hello @chapman91, Make your importation of `AggregatorV3Interface` be from the same source in both your `PriceConverter` Library and your `FundMe` contract

**Reply by chapman91** [reply]
Hey @EngrPips, I checked and its verified that the AgggregatorV3Interface both in the Price Converter Library and FundMe contract are coming from the same place.

**Reply by EngrPips** [reply]
> Hey @EngrPips, I checked and its verified that the AgggregatorV3Interface both in the Price Converter Library and FundMe contract are coming from the same place.

Please share your `FundMe` contract and your `PriceConverter` Library, Please don't send images but copy the code and send them also make sure to format them appropriately.

**Comment by chapman91** [comment]
### **Below you have the Fundme contract**

`// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error FundMe__NotOwner();

contract FundMe {
    using PriceConverter for uint256;

    mapping(address => uint256) public addressToAmountFunded;
    address[] public funders;

    address public /* immutable */ i_owner;
     
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;  // ** acts as the exponentiation operator
   
    AggregatorV3Interface private s_priceFeed;



    constructor(address priceFeed) {
        s_priceFeed = AggregatorV3Interface(priceFeed);
        i_owner = msg.sender;

    }

    function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }
    // getVersion Function
    function getVersion() public view returns (uint256) {
        return s_priceFeed.version();
    }

    modifier onlyOwner() {
        if (msg.sender != i_owner) revert FundMe__NotOwner();
        _;
    }

    function withdraw() public onlyOwner {
        for (uint256 funderIndex = 0; funderIndex < funders.length; funderIndex++) {
            address funder = funders[funderIndex];
            addressToAmountFunded[funder] = 0;
        }
        funders = new address[](0);


        (bool callSuccess,) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }

    fallback() external payable {
        fund();
    }

    receive() external payable {
        fund();
    }
}



### **Below is the PriceConverter Contract**

`// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

library PriceConverter {
    function getPrice(
        AggregatorV3Interface priceFeed
    ) internal view returns (uint256) {

        (, int256 answer, , , ) = priceFeed.latestRoundData();
        // ETH/USD rate in 18 digit
        return uint256(answer * 10000000000);
    }

    function getConversionRate(
        uint256 ethAmount, // Amount of Ether (in wei) to be converteed
        AggregatorV3Interface priceFeed // Price feed interface to get the current ETH/USD price
    ) internal view returns (uint256) {
        uint256 ethPrice = getPrice(priceFeed); // Get the currenct ETH price in USD (with 18 decimals)
        uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000; // calculate the equivalent USD value
        // the actual ETH/USD conversion rate, after adjusting the extra 0s.
        return ethAmountInUsd;
    }
}
`

**Comment by EngrPips** [comment]
**FUNDME CONTRACT**

```js

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error FundMe__NotOwner();

contract FundMe {
using PriceConverter for uint256;

mapping(address => uint256) public addressToAmountFunded;
address[] public funders;

address public /* immutable */ i_owner;
 
uint256 public constant MINIMUM_USD = 5 * 10 ** 18;  // ** acts as the exponentiation operator

AggregatorV3Interface private s_priceFeed;



constructor(address priceFeed) {
    s_priceFeed = AggregatorV3Interface(priceFeed);
    i_owner = msg.sender;

}

function fund() public payable {
    require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
    // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
    addressToAmountFunded[msg.sender] += msg.value;
    funders.push(msg.sender);
}
// getVersion Function
function getVersion() public view returns (uint256) {
    return s_priceFeed.version();
}

modifier onlyOwner() {
    if (msg.sender != i_owner) revert FundMe__NotOwner();
    _;
}

function withdraw() public onlyOwner {
    for (uint256 funderIndex = 0; funderIndex < funders.length; funderIndex++) {
        address funder = funders[funderIndex];
        addressToAmountFunded[funder] = 0;
    }
    funders = new address[](0);


    (bool callSuccess,) = payable(msg.sender).call{value: address(this).balance}("");
    require(callSuccess, "Call failed");
}

fallback() external payable {
    fund();
}

receive() external payable {
    fund();
}
}
```

**PRICE CONVERTER LIBRARY**

```js

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

library PriceConverter {
function getPrice(
AggregatorV3Interface priceFeed
) internal view returns (uint256) {

    (, int256 answer, , , ) = priceFeed.latestRoundData();
    // ETH/USD rate in 18 digit
    return uint256(answer * 10000000000);
}

function getConversionRate(
    uint256 ethAmount, // Amount of Ether (in wei) to be converteed
    AggregatorV3Interface priceFeed // Price feed interface to get the current ETH/USD price
) internal view returns (uint256) {
    uint256 ethPrice = getPrice(priceFeed); // Get the currenct ETH price in USD (with 18 decimals)
    uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000; // calculate the equivalent USD value
    // the actual ETH/USD conversion rate, after adjusting the extra 0s.
    return ethAmountInUsd;
}
}
```

**Reply by EngrPips** [reply, marked as answer]
The path you imported `AggregatorV3Interface` from in your `FundMe` contract is `import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";`. In contrast, the path you imported `AggregatorV3Interface` from in your `PriceConverter` Library is `import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";` They are not the same, Please kindly make use of one path in both the `FundMe` contract and `PriceConverter` Library.

**Reply by chapman91** [reply]
Thank you. I bypassed the error. Learnt a lesson to look closer next time!

**Reply by EngrPips** [reply]
Indeed, Programming is always teaching us a lot.
